proto/api.proto

```diff
@@ -135,27 +119,53 @@ message ResponseHeader {
   optional Transaction txn = 3;
 }
 
+// KVRequestHeader is supplied with every KV node request.
+message KVRequestHeader {
+  optional RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  // The key for request. If the request operates on a range, this
+  // represents the starting key for the range.
+  optional bytes key = 3 [(gogoproto.casttype) = "Key"];
+  // The end key is empty if the request spans only a single key. Otherwise,
+  // it must order strictly after Key. In such a case, the header indicates
+  // that the operation takes place on the key range from Key to EndKey,
+  // including Key and excluding EndKey.
+  optional bytes end_key = 4 [(gogoproto.casttype) = "Key"];
+  // Replica specifies the destination for the request. This is a specific
+  // instance of the available replicas belonging to RangeID.
+  optional Replica replica = 6 [(gogoproto.nullable) = false];
+  // RaftID specifies the ID of the Raft consensus group which the key
+  // range belongs to. This is used by the receiving node to route the
+  // request to the correct range.
+  optional int64 raft_id = 7 [(gogoproto.nullable) = false,
+      (gogoproto.customname) = "RaftID", (gogoproto.casttype) = "RaftID"];
+}
+
+// KVResponseHeader is supplied with every KV node request.
+message KVResponseHeader {
+  optional ResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+}
+
 // A GetRequest is the argument for the Get() method.
 message GetRequest {
-  optional RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional KVRequestHeader kvheader = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
```

Comment:
I dislike relying on the embed option here. In go, we can use request.Header() and pretend the intermediate KVHeader doesn't exist, but in other languages you'll have to go through the intermediate `{KV,Structured}Header`, and that will make it more difficult to deal with all requests in a generic way. I would prefer to have two separate top-level fields for RequestHeader and KVRequestHeader.